File: src/main/java/com/winter/mayawinterfox/command/Command.java
package com.winter.mayawinterfox.command;

import discord4j.core.event.domain.message.MessageCreateEvent;

import java.util.HashSet;
import java.util.Set;
import java.util.function.Predicate;

public class Command {

	private final String name;
	private final String help;
	private final Predicate<MessageCreateEvent> check;
	private final Predicate<MessageCreateEvent> call;
	private final Set<String> aliases;

	/**
	 * Creates a new command class
	 * @param name  Name of the command or sub-command.
	 * @param help  The help message key of the command.
	 * @param check Whether to run the call or not.
	 * @param call  Run the call, false means show help message. Help is determined by <code>[name]-help</code>.
	 */
	public Command(String name, String help, Predicate<MessageCreateEvent> check, Predicate<MessageCreateEvent> call) {
		this(name, help, check, call, new HashSet<>());
	}

	/**
	 * Creates a new command class
	 * @param name  Name of the command or sub-command.
	 * @param help  The help message key of the command.
	 * @param check Whether to run the call or not.
	 * @param call  Run the call, false means show help message. Help is determined by <code>[name]-help</code>.
	 * @param aliases The aliases for the command
	 */
	public Command(String name, String help, Predicate<MessageCreateEvent> check, Predicate<MessageCreateEvent> call, Set<String> aliases) {
		this.name = name;
		this.help = help;
		this.check = check;
		this.call = call;
		this.aliases = new HashSet<>(aliases);
		this.aliases.add(name);
	}

	public void call(MessageCreateEvent e) {
		call.test(e);
	}

	public String getName() {
		return name;
	}

	public String getHelp() {
		return help;
	}

	public Predicate<MessageCreateEvent> getCheck() { return check; }

	public Set<String> getAliases() {
		return aliases;
	}

}

File: src/main/java/com/winter/mayawinterfox/command/Command.kt
package com.winter.mayawinterfox.command;

import com.winter.mayawinterfox.command.Command.Category
import discord4j.core.event.domain.message.MessageCreateEvent
import reactor.core.publisher.Mono

abstract class Command(
		/**
		 * Name of the command
		 */
		val name: String,

		/**
		 * The help message key in the localization file for the necessary translations
		 */
		val help: String,

		/**
		 * The category of the command
		 * @see Category
		 */
		val category: Category,

		/**
		 * The permission command for this command, checks bot & user permissions
		 * before execution of the command
		 * @see CommandPermission
		 */
		val permissions: CommandPermission,

		/**
		 * Aliases for the command
		 */
		val aliases: Set<String> = emptySet()
) {
	/**
	 * Execute the command, this does not do permission checking
	 */
	abstract fun call(event: MessageCreateEvent): Mono<Void>

	/**
	 * All the command categories and their labels
	 */
	enum class Category(val type: String) {
		ADMIN("admin"),
		FUN("fun"),
		DEV("developer"),
		STATUS("status"),
		MISC("miscellaneous"),
		ANIMAL("animal"),
		PROFILE("profile"),
		UTIL("utility"),
		MUSIC("music"),
		IMAGE("image")
	}
}